src/IR/LowerAst.hs

```diff
@@ -48,10 +48,47 @@ lowerType a@(A.TApp _ _      ) = case A.collectTApp a of
 
 -- | Lower an AST 'Program' into IR.
 lowerProgram :: A.Program -> Compiler.Pass (I.Program I.Type)
-lowerProgram (A.Program ds) = return $ I.Program
+lowerProgram (A.Program []) = return $ I.Program
   { I.programEntry = fromString "main"
-  , I.programDefs  = map (first fromJust . lowerDef) ds
+  , I.programDefs  = []
   , I.typeDefs     = []
+  , I.classDefs    = []
+  , I.instDefs     = []
+  }
+lowerProgram (A.Program (A.TopDef d : tds)) =
```

Comment:
This should be broken up into some kind of map or fold; the amount of code here suggests that this is doing more than it is.

My suggestion is to filter the definition list into three different lists, of defs, classes, and insts, and then map `lower{Def,Class,Inst}` over those. Something like:

```haskell
return $ I.Program
  { I.programEntry = fromString "main"
  , I.programDefs = map lowerTopDef $ filterBy (\case -> { A.TopDef d -> Just d; _ -> Nothing }) ds
  , I.classDefs = map lowerTopDef $ filterBy (\case -> { A.TopClass c -> Just c; _ -> Nothing }) ds
  , I.programDefs = map lowerTopDef $ filterBy (\case -> { A.TopInst i -> Just i; _ -> Nothing }) ds
  -- etc.
  }
  where filterBy f = catMaybes . map f
```